### PR TITLE
package: Add support installation of Preinstalled firmware.

### PR DESCRIPTION
**Pull request by Zangetsu38**
# About
- package: Add support installation of Preinstalled firmware. 
Add support of exfat partition for installing it.
- gui/initial setup: Add button for download Preinstalled Firmware.

Get usefull for later with vsh and also with my other pr of bgm support

### CODE REVIEW

**Comment by ArbestRi02** [comment]
Getting the VSH to run will be insane (but we already did that a year ago with the Bankai PR)

**Review by bookmist — DISMISSED** [review]

File: vita3k/packages/src/exfat.cpp [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+            utf8String.push_back(static_cast<char>(ch));
+        } else if (ch < 0x800) {
+            utf8String.push_back(0xC0 | (ch >> 6));
+            utf8String.push_back(0x80 | (ch & 0x3F));
+        } else {
+            utf8String.push_back(0xE0 | (ch >> 12));
+            utf8String.push_back(0x80 | ((ch >> 6) & 0x3F));
+            utf8String.push_back(0x80 | (ch & 0x3F));
+        }
+    }
```
Comment (bookmist) [inline vita3k/packages/src/exfat.cpp:61]: here `string_utils::wide_to_utf` can be used

File: vita3k/packages/src/exfat.cpp [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        case EXFAT_ENTRY_FILE: {
+            // Read the file info
+            ExFATFileEntryInfo file_info;
+            img.read(reinterpret_cast<char *>(&file_info), sizeof(ExFATFileEntryInfo));
+
+            // Get the name of file or directory
+            std::string name = get_exfat_file_name(img, file_entry.continuations);
+
+            // Set path of the current file or directory
+            const auto subdir = current_dir / name;
```
Comment (bookmist) [inline vita3k/packages/src/exfat.cpp:101]:
```suggestion
            const auto subdir = current_dir / fs_utils::utf8_to_path(name);
```

File: vita3k/packages/include/packages/sce_types.h [hunk 3]
```diff
@@ -651,6 +651,7 @@ class SceRIF {
 };
 
 void register_keys(KeyStore &SCE_KEYS, int type);
+void extract_exfat(const fs::path &partition_path, const std::string &partition, const fs::path &pref_path);
```
Comment (bookmist) [inline vita3k/packages/include/packages/sce_types.h:654]: Why it's here? This function already defined in `packages/exfat.h`

Comment (Zangetsu38) [reply vita3k/packages/include/packages/sce_types.h:654]: Oupsy 😐

File: vita3k/packages/src/exfat.cpp (hunk 1 above)
Comment (Zangetsu38) [reply vita3k/packages/src/exfat.cpp:61]: It can convert utf16 to utf8 ?

Comment (Zangetsu38) [reply vita3k/packages/src/exfat.cpp:61]: > here `string_utils::wide_to_utf` can be used

how ? this functions request wstring, and here is uint16_t so i no can using it

Comment (bookmist) [reply vita3k/packages/src/exfat.cpp:61]: Ok. This is not really important. Just a possibility.

Comment (Zangetsu38) [reply vita3k/packages/src/exfat.cpp:61]: Forget, i have succes lol

**Review by bookmist — DISMISSED** [review]

**Pull request merged**